### PR TITLE
Issue #37: Data Binding on Characters Activity

### PR DESCRIPTION
**Pull request by jamacado**
**Data Binding on Characters Activity** #37

-  Delete charactersUiModel because implemented databinding.
- Create diferent bindingAdapter for use in xml.
- Create Event for use navigation.

**Testing Done:**

- Nexus 6P

**How to test:**

- Run the application and click onCharacter Category
- You can see the loading while the characters have been loaded
- Then you can see the characters in a scrollable list

**Signed-off-by**:

- Javier Cardona <javier.mcardona@gmail.com>

### CODE REVIEW

File: presentation/src/main/java/com/architect/coders/mu8/characters/CharactersActivity.kt [hunk 1]
```diff
@@ -1,46 +1,32 @@
...
         super.onCreate(savedInstanceState)
-        setContentView(R.layout.activity_characters)
-
-        setSupportActionBar(toolbar)
-        supportActionBar?.setDisplayShowTitleEnabled(false)
 
-        val toolbarTitle = findViewById<TextView>(R.id.toolbar_title)
-        toolbarTitle.text = getString(R.string.characters_name)
+        val binding: ActivityCharactersBinding =
+            DataBindingUtil.setContentView(this, R.layout.activity_characters)
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/characters/CharactersActivity.kt]: Put this in a single line.

File: presentation/src/main/java/com/architect/coders/mu8/characters/CharactersActivity.kt [hunk 2]
```diff
@@ -50,22 +36,28 @@ class CharactersActivity : AppCompatActivity() {
...
-        adapter = CharactersAdapter(viewModel::onCharacterClicked)
-        recycler.layoutManager = LinearLayoutManager(this)
-        recycler.adapter = adapter
+        binding.viewmodel = viewModel
+        binding.lifecycleOwner = this
 
-        viewModel.model.observe(this, Observer(::updateUi))
-    }
-
-    private fun updateUi(model: CharactersUiModel) {
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/characters/CharactersActivity.kt:52]: Why did you leave everything inside the `onCreate` method? It looks big, please create smaller methods for specific initializations and instead call them here.

File: presentation/src/main/java/com/architect/coders/mu8/characters/CharactersBindingAdapters.kt [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.architect.coders.mu8.characters
+
+import androidx.databinding.BindingAdapter
+import androidx.recyclerview.widget.RecyclerView
+import com.architect.codes.mu8.characters.Character
+
+@BindingAdapter("items")
+fun RecyclerView.setItems(characters: List<Character>?) {
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/characters/CharactersBindingAdapters.kt]: Didn't we said we were going to remove this type of methods?

File: presentation/src/main/java/com/architect/coders/mu8/utils/BindingAdapters.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package com.architect.coders.mu8.utils
+
+import android.view.View
+import androidx.databinding.BindingAdapter
+
+@BindingAdapter("visible")
+fun View.setVisible(visible: Boolean?) {
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/utils/BindingAdapters.kt:7]: Same comment about the functions, can we avoid it? I'll double check Antonio's video tho.

File: presentation/src/main/res/layout/view_character.xml [hunk 5]
```diff
@@ -1,31 +1,40 @@
...
-        android:contentDescription="@null"
-        app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintStart_toStartOf="parent"
-        app:layout_constraintTop_toTopOf="parent" />
+        android:layout_height="wrap_content">
+
+        <ImageView
+            android:id="@+id/image_character_picture"
+            android:layout_width="match_parent"
+            android:layout_height="200dp"
```
Comment (recaldev) [inline presentation/src/main/res/layout/view_character.xml]: Use dimens

File: presentation/src/main/res/layout/view_character.xml [hunk 6]
```diff
@@ -1,31 +1,40 @@
...
-        app:layout_constraintBottom_toBottomOf="@+id/image_character_picture"
-        app:layout_constraintEnd_toEndOf="@+id/image_character_picture"
-        tools:text="Batman Infiltrado" />
-</androidx.constraintlayout.widget.ConstraintLayout>
\ No newline at end of file
+        <TextView
+            android:id="@+id/text_character_name"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginEnd="10dp"
```
Comment (recaldev) [inline presentation/src/main/res/layout/view_character.xml]: Use dimens

File: presentation/src/main/res/layout/view_character.xml [hunk 7]
```diff
@@ -1,31 +1,40 @@
...
+            android:layout_marginBottom="10dp"
```
Comment (recaldev) [inline presentation/src/main/res/layout/view_character.xml]: Use dimens

File: presentation/src/main/res/layout/view_character.xml [hunk 8]
```diff
@@ -1,31 +1,40 @@
...
+            android:elevation="5dp"
```
Comment (recaldev) [inline presentation/src/main/res/layout/view_character.xml]: Use dimens

File: presentation/src/main/res/layout/view_character.xml [hunk 9]
```diff
@@ -1,31 +1,40 @@
...
+            android:shadowColor="@android:color/black"
+            android:shadowRadius="50"
```
Comment (recaldev) [inline presentation/src/main/res/layout/view_character.xml]: Use dimens

**Review by recaldev — CHANGES_REQUESTED** [review]

File: presentation/src/main/java/com/architect/coders/mu8/characters/CharactersAdapter.kt [hunk 10]
```diff
@@ -6,6 +6,8 @@ import android.widget.ImageView
 import android.widget.TextView
```
Comment (recaldev) [inline presentation/src/main/java/com/architect/coders/mu8/characters/CharactersAdapter.kt]: Please review the imports on all classes, this one in particular has several unused imports.

File: presentation/src/main/java/com/architect/coders/mu8/characters/CharactersActivity.kt (hunk 1 above)
Comment (jamacado) [reply presentation/src/main/java/com/architect/coders/mu8/characters/CharactersActivity.kt]: ok,

File: presentation/src/main/java/com/architect/coders/mu8/characters/CharactersBindingAdapters.kt (hunk 3 above)
Comment (jamacado) [reply presentation/src/main/java/com/architect/coders/mu8/characters/CharactersBindingAdapters.kt]: ok, delete file

File: presentation/src/main/res/layout/view_character.xml (hunk 5 above)
Comment (jamacado) [reply presentation/src/main/res/layout/view_character.xml]: create dp_200 and change value for dimen

File: presentation/src/main/res/layout/view_character.xml (hunk 7 above)
Comment (jamacado) [reply presentation/src/main/res/layout/view_character.xml]: change for dimen dp_10

File: presentation/src/main/res/layout/view_character.xml (hunk 8 above)
Comment (jamacado) [reply presentation/src/main/res/layout/view_character.xml]: change dp_5

File: presentation/src/main/res/layout/view_character.xml (hunk 9 above)
Comment (jamacado) [reply presentation/src/main/res/layout/view_character.xml]: change dp_50

File: presentation/src/main/java/com/architect/coders/mu8/characters/CharactersAdapter.kt (hunk 10 above)
Comment (jamacado) [reply presentation/src/main/java/com/architect/coders/mu8/characters/CharactersAdapter.kt]: ok

File: data/src/main/java/com/architect/coders/mu8/data/database/MU8Database.kt [hunk 11]
```diff
@@ -9,7 +9,7 @@ import com.architect.coders.mu8.data.database.converter.ListStringTypeConverters
 import com.architect.coders.mu8.data.database.converter.ThumbnailTypeConverters
 import com.architect.coders.mu8.data.database.converter.UrlsTypeConverters
 
-private const val DATABASE_VERSION = 1
+private const val DATABASE_VERSION = 2
```
Comment (aliceresponde) [inline data/src/main/java/com/architect/coders/mu8/data/database/MU8Database.kt:12]: Why  did you  upgrade  the  database version?

File: presentation/src/main/java/com/architect/coders/mu8/characters/CharactersBindingAdapters.kt (hunk 3 above)
Comment (recaldev) [reply presentation/src/main/java/com/architect/coders/mu8/characters/CharactersBindingAdapters.kt]: Esto vuela

**Review by recaldev — APPROVED** [review]

**Review by aliceresponde — APPROVED** [review]
LG

**Pull request merged**